### ISSUE TITLE
git always pulls from master instead from tracking branch

### ISSUE THREAD

**Issue by jlec** [ansible/ansible#19139]
_From @jlec on August 21, 2015 10:7_

Issue Type: Bug report
Ansible Version: 1.9.2
Ansible Configuration: ?
Environment: Gentoo Linux 64bit
Summary: remote checkout tracks branch develop but git module always updates from master
Steps To Reproduce:
- Clone remote repo onto branch != master, with master/HEAD != branch/HEAD
- ....  -m git -a "repo=git://..../foo.git dest=.../foo update=yes"
- check repo status
  Expected Results: pull from tracking branch

Actual Results:
On remote:

```
 git branch -av
* develop                dc66049 Basic Celery support
  remotes/origin/HEAD    -> origin/master
  remotes/origin/develop dc66049 Basic Celery support
  remotes/origin/master  049a7c5 Merge branch 'release/0.4.2'
```

After running
`ansible foo-staging -m git -a "repo=git://.../foo.git dest=../foo update=yes"`

```
# git branch -av
* develop                049a7c5 [behind 4] Merge branch 'release/0.4.2'
  remotes/origin/HEAD    -> origin/master
  remotes/origin/develop dc66049 Basic Celery support
  remotes/origin/master  049a7c5 Merge branch 'release/0.4.2'
```

_Copied from original issue: ansible/ansible-modules-core#1948_

**Comment by jlec** [comment]
_From @robinro on April 18, 2016 12:39_

You need version=develop to specify the branch you want.

If you want it to respect the currently checkout out branch, that would be a quite different feature.

**Comment by jlec** [comment]
_From @ansibot on August 1, 2016 2:49_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jlec** [comment]
_From @ansibot on September 12, 2016 21:1_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jlec** [comment]
@robinro I don't want to specify a special branch and this has nothing to do with the checked out branch on the remote. I would like to use the default remote tracking branch of the git repo. This normally is master, but could be different. Ansible always uses master.

**Comment by jlec** [comment]
_From @robinro on September 14, 2016 16:19_

@jlec 
When you clone a repo, the branch that is checked out is the one that `HEAD` points to, see e.g. also http://stackoverflow.com/questions/18726037/what-determines-default-branch-after-git-clone
This is independent of the checkout.

You can specify this also explicitly by passing `version=HEAD` to the ansible module.

If this doesn't clarify the issue, please provide a public example repository, where a shell command `git clone` and the ansible git module lead to different results.

**Comment by jlec** [comment]
_From @ansibot on October 6, 2016 15:50_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jlec** [comment]
_From @ansibot on October 22, 2016 9:19_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jlec** [comment]
_From @ansibot on November 6, 2016 10:9_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jlec** [comment]
See the following example

Check out the repo remotely using

`ansible all -i inventory -m git -a "repo=https://github.com/ansible/ansible-modules-core.git dest=/tmp/git-test update=yes"`

So we are on branch `devel`tracking `origin/devel`

Then log in into the remote box and run

`git checkout -b staging origin/staging``

Now we are on branch `staging`tracking `origin/staging`.

```
$ git branch
  devel     931a713 If fetch_url failed to download the URL fail early with a proper error message. Fixes #5474 (#5476)
* staging 26636bb Improved output of mysql_db plugin
```

When running the original ansible command again

`ansible all -i inventory -m git -a "repo=https://github.com/ansible/ansible-modules-core.git dest=/tmp/git-test update=yes"`

we end up with 

```
$ git branch
  devel   931a713 If fetch_url failed to download the URL fail early with a proper error message. Fixes #5474 (#5476)
* staging 931a713 [ahead 3556, behind 7] If fetch_url failed to download the URL fail early with a proper error message. Fixes #5474 (#5476)
```

This means, instead of respecting the repo's tracking settings, the `origin/HEAD`got pulled.

Although you are right, setting `version`would solve the problem, I would intuitively assume, tracking is respected.

**Comment by jlec** [comment]
_From @ansibot on November 22, 2016 10:17_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jlec** [comment]
_From @ansibot on December 9, 2016 19:54_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
@jlec Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jlec** [comment]
##### COMPONENT NAME
git

**Comment by robinro** [comment]
As stated in my earlier comment I believe the ansible module behaves exactly like native git, i.e. follows the remote HEAD when no version is specified.
close_me

If I misunderstood, please provide a public example repository, where a shell command `git clone` and the ansible git module lead to different results.

**Comment by ansibotdev** [comment]
@jlec You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
